tests/unittests/media/client/main/mediaPipeline/MediaPipelineProxyTest.cpp

```diff
@@ -0,0 +1,142 @@
+
+    std::shared_ptr<MediaPipelineProxy> proxy;
+    EXPECT_NO_THROW(proxy = std::make_shared<MediaPipelineProxy>(mediaPipelineMock, *m_clientControllerMock));
+
+    const std::string kMimeType{"mime"};
+    const std::string kUrl{"url"};
+    const std::unique_ptr<firebolt::rialto::IMediaPipeline::MediaSource> kMediaSource;
+    const int kSourceId{2};
+    const double kPlaybackRate{2.1};
+    const double kVolume1{3.1};
+    const double kVolume2{4.1};
+    const int64_t kPosition1{123};
+    const int64_t kPosition2{234};
+    const uint32_t kNeedDataRequestId{5};
+    const std::unique_ptr<firebolt::rialto::IMediaPipeline::MediaSegment> kMediaSegment;
+    const std::shared_ptr<IMediaPipelineClient> kIMediaPipelineClient;
+
+    /////////////////////////////////////////////
+
+    EXPECT_CALL(*mediaPipelineMock, load(MediaType::MSE, StrEq(kMimeType), StrEq(kUrl))).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, attachSource(_)).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, removeSource(kSourceId)).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, allSourcesAttached()).WillOnce(Return(true));
+
+    EXPECT_CALL(*mediaPipelineMock, play()).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, pause()).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, stop()).WillOnce(Return(true));
+
+    EXPECT_CALL(*mediaPipelineMock, setPlaybackRate(DoubleEq(kPlaybackRate))).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, setPosition(kPosition1)).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, getPosition(_)).WillOnce(DoAll(SetArgReferee<0>(kPosition2), Return(true)));
+    EXPECT_CALL(*mediaPipelineMock, setVideoWindow(1, 2, 3, 4)).WillOnce(Return(true));
+
+    EXPECT_CALL(*mediaPipelineMock, haveData(MediaSourceStatus::CODEC_CHANGED, kNeedDataRequestId)).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, addSegment(kNeedDataRequestId, _)).WillOnce(Return(AddSegmentStatus::OK));
+    EXPECT_CALL(*mediaPipelineMock, renderFrame()).WillOnce(Return(true));
+
+    EXPECT_CALL(*mediaPipelineMock, setVolume(DoubleEq(kVolume1))).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, getVolume(_)).WillOnce(DoAll(SetArgReferee<0>(kVolume2), Return(true)));
+    EXPECT_CALL(*mediaPipelineMock, setMute(true)).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, getMute(_)).WillOnce(DoAll(SetArgReferee<0>(false), Return(true)));
+
+    EXPECT_CALL(*mediaPipelineMock, flush(kSourceId, true)).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, setSourcePosition(kSourceId, kPosition1)).WillOnce(Return(true));
+    EXPECT_CALL(*mediaPipelineMock, getClient()).WillOnce(Return(kIMediaPipelineClient));
+    EXPECT_CALL(*mediaPipelineMock, notifyApplicationState(ApplicationState::RUNNING));
+
+    /////////////////////////////////////////////
+
+    EXPECT_TRUE(proxy->load(MediaType::MSE, kMimeType, kUrl));
```

Comment:
I would prefer the EXPECT_CALL to be set just before we call the method, e.g.:

EXPECT_CALL(*mediaPipelineMock, attachSource(_)).WillOnce(Return(true));
EXPECT_TRUE(proxy->attachSource(kMediaSource));

For example what is proxy->play() actually called media_pipeline->pause() for some strange reason. EXPECT_TRUE(proxy->play()); would still succeed because we have set EXPECT_CALL(*mediaPipelineMock, pause()).WillOnce(Return(true)); prior to this. EXPECT_TRUE(proxy->pause()); would then fail even though it behaved correctly.